lib/services/readonly/readonly.go

```diff
@@ -0,0 +1,105 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package readonly
+
+import (
+	"time"
+
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/keys"
+)
+
+// NOTE: is best to avoid importing anything from lib other than lib/utils in this package in
+// order to ensure that we can import it anyhwere api/types is being used.
+
+// AuthPreference is a read-only subset of types.AuthPreference used on certain hot paths
+// to ensure that we do not modify the underlying AuthPreference as it may be shared across
+// multiple goroutines.
+type AuthPreference interface {
+	GetSecondFactor() constants.SecondFactorType
+	GetDisconnectExpiredCert() bool
+	GetLockingMode() constants.LockingMode
+	GetDeviceTrust() *types.DeviceTrust
+	GetPrivateKeyPolicy() keys.PrivateKeyPolicy
+	IsAdminActionMFAEnforced() bool
+	GetRequireMFAType() types.RequireMFAType
+	IsSAMLIdPEnabled() bool
+	GetDefaultSessionTTL() types.Duration
+	GetHardwareKeySerialNumberValidation() (*types.HardwareKeySerialNumberValidation, error)
+	GetAllowPasswordless() bool
+	Clone() types.AuthPreference
+}
+
+type sealedAuthPreference struct {
+	AuthPreference
+}
+
+// SealAuthPreference returns a read-only version of the AuthPreference.
+func SealAuthPreference(p types.AuthPreference) AuthPreference {
```

Comment:
do we need to export the Seal functions?

Reply:
Accidental holdover from an earlier iteration of this feature that used sealing in multiple packages.  Removed.